### PR TITLE
add hotkey trim trigger function

### PR DESCRIPTION
**Pull request by xooiamd**
- Add hotkey trim trigger for Windows and Linux Desktop platform. 
- The hotkey trim trigger can be enabled by specifying a hotkey (any one of F1-F12|TAB|CONTROL) in the environment GFXRECON_TRIM_TRIGGER, e.g. set GFXRECON_TRIM_TRIGGER=F3 will enable trim trigger start/stop capture by pressing hotkey F3.

### CODE REVIEW

File: framework/encode/capture_settings.cpp [hunk 1]
```diff
@@ -75,6 +76,8 @@ GFXRECON_BEGIN_NAMESPACE(encode)
 #define PAGE_GUARD_SEPARATE_READ_UPPER      "PAGE_GUARD_SEPARATE_READ"
 #define PAGE_GUARD_EXTERNAL_MEMORY_LOWER    "page_guard_external_memory"
 #define PAGE_GUARD_EXTERNAL_MEMORY_UPPER    "PAGE_GUARD_EXTERNAL_MEMORY"
+#define TRIM_TRIGGER_LOWER                  "trim_trigger"
```
[the remark this answers is not in the document]
Comment (xooiamd) [reply framework/encode/capture_settings.cpp]: changed to GFXRECON_CAPTURE_FRAMES

File: framework/encode/trace_manager.cpp [hunk 2]
```diff
@@ -376,48 +409,220 @@ void TraceManager::EndApiCallTrace(ParameterEncoder* encoder)
...
 
-void TraceManager::EndFrame()
+#if defined(__linux__)
+static xcb_connection_t* keyboard_connection = nullptr;
+
+// On Linux platform, xcb calls need Connection which is connected to target
+// server, because hotkey process is supposed to insert into target application,
+// so we need to capture the connection that target app use. the function is
+// used to insert into Vulkan call to capture and save the connection.
+// TODO(xooi@amd.com) insert this function into Vulkan vkCreateXcbSurfaceKHR call.
```
[the remark this answers is not in the document]
Comment (xooiamd) [reply framework/encode/trace_manager.cpp]: sorry the TODO is done so i removed them,

File: framework/encode/trace_manager.cpp [hunk 3]
```diff
@@ -219,7 +227,7 @@ bool TraceManager::Initialize(std::string base_filename, const CaptureSettings::
         page_guard_external_memory_ = false;
     }
 
-    if (trace_settings.trim_ranges.empty())
+    if (trace_settings.trim_ranges.empty() && trace_settings.trim_key == "")
```
[the remark this answers is not in the document]
Comment (xooiamd) [reply framework/encode/trace_manager.cpp]: change to .empty()

File: framework/encode/trace_manager.cpp [hunk 4]
```diff
@@ -29,6 +30,13 @@
 
 #include <cassert>
 
+#if defined(__linux__)
```
[the remark this answers is not in the document]
Comment (xooiamd) [reply framework/encode/trace_manager.cpp]: changed to #if defined(__linux__) && !defined(__ANDROID__)

File: framework/encode/trace_manager.cpp [hunk 5]
```diff
@@ -29,6 +30,10 @@
...
+#include <xcb/xcb_keysyms.h>
```
[the remark this answers is not in the document]
Comment (xooiamd) [reply framework/encode/trace_manager.cpp]: updated BUILD.md

File: framework/encode/trace_manager.cpp [hunk 6]
```diff
@@ -230,17 +237,42 @@ bool TraceManager::Initialize(std::string base_filename, const CaptureSettings::
...
+
+                success = CreateCaptureFile(CreateTrimFilename(base_filename_, trim_ranges_[0]));
             }
+            else
+            {
+                capture_mode_ = kModeTrack;
+            }
+        }
+        // Check if trim is enabled by hot-key trigger at the first frame
+        else if (trace_settings.trim_key != "")
```
[the remark this answers is not in the document]
Comment (xooiamd) [reply framework/encode/trace_manager.cpp]: changed.

File: framework/encode/trace_manager.cpp [hunk 7]
```diff
@@ -230,17 +237,42 @@ bool TraceManager::Initialize(std::string base_filename, const CaptureSettings::
...
+        {
+            trim_key_ = trace_settings.trim_key;
+
+            // Enable state tracking when hotkey pressed
+            if (IsTrimHotkeyPressed() == true)
```
[the remark this answers is not in the document]
Comment (xooiamd) [reply framework/encode/trace_manager.cpp]: remove the explicit true conditions

File: framework/encode/trace_manager.cpp [hunk 8]
```diff
@@ -376,48 +408,232 @@ void TraceManager::EndApiCallTrace(ParameterEncoder* encoder)
     }
 }
 
-void TraceManager::EndFrame()
+#if defined(__linux__)
```
[the remark this answers is not in the document]
Comment (xooiamd) [reply framework/encode/trace_manager.cpp]: add #ifdef VK_USE_PLATFORM_XCB_KHR

File: framework/encode/trace_manager.cpp [hunk 9]
```diff
@@ -376,48 +408,232 @@ void TraceManager::EndApiCallTrace(ParameterEncoder* encoder)
...
+                                                                 { "ControlLeft", XK_Control_L },
+                                                                 { "ControlRight", XK_Control_R } };
+
+#elif defined(__ANDROID__)
+    // TODO(xooi@amd.com) implement key_code_map for Android
+    static std::unordered_map<std::string, int> key_code_map = {};
+#endif
+
+    bool                                           hotkey_triggered  = false;
+    std::unordered_map<std::string, int>::iterator iterator_key_code = key_code_map.find(trim_key_);
```
[the remark this answers is not in the document]
Comment (xooiamd) [reply framework/encode/trace_manager.cpp]: changed to auto

File: framework/encode/capture_settings.h [hunk 10]
```diff
@@ -62,6 +63,7 @@ class CaptureSettings
         bool                   force_flush{ false };
         MemoryTrackingMode     memory_tracking_mode{ kPageGuard };
         std::vector<TrimRange> trim_ranges;
+        std::string            trim_key{ "" };
```
[the remark this answers is not in the document]
Comment (xooiamd) [reply framework/encode/capture_settings.h]: Done.

File: framework/encode/trace_manager.h [hunk 11]
```diff
@@ -858,6 +870,7 @@ class TraceManager
     static thread_local std::unique_ptr<ThreadData> thread_data_;
     static LayerTable                               layer_table_;
     static std::atomic<format::HandleId>            unique_id_counter_;
+    static bool                                     previous_hotkey_trigger;
```
[the remark this answers is not in the document]
Comment (xooiamd) [reply framework/encode/trace_manager.h]: Done.

File: framework/encode/capture_settings.cpp [hunk 12]
```diff
@@ -589,5 +611,21 @@ void CaptureSettings::ParseTrimRangeString(const std::string&
     }
 }
 
+std::string CaptureSettings::ParseTrimKeyString(const std::string& value_string)
+{
+    std::string trim_key = "";
```
[the remark this answers is not in the document]
Comment (xooiamd) [reply framework/encode/capture_settings.cpp]: Done.

File: framework/encode/trace_manager.cpp [hunk 13]
```diff
@@ -376,48 +408,232 @@ void TraceManager::EndApiCallTrace(ParameterEncoder* encoder)
...
         }
-        else if ((capture_mode_ & kModeTrack) == kModeTrack)
+    }
+    else if (IsTrimHotkeyPressed() == true)
+    {
+        // Stop recording and close file.
+        capture_mode_ &= ~kModeWrite;
+        file_stream_ = nullptr;
+        GFXRECON_LOG_INFO("Finished recording graphics API capture");
+        // Capture can be disabled and resources can be released.
```
[the remark this answers is not in the document]
Comment (xooiamd) [reply framework/encode/trace_manager.cpp]: Done.

**Comment by simonvanderveldt** [comment]
Driveby comment, and I probably won't have time to test this in the future but this didn't work for me. 
Logs showed the setting was picked up but pushing F12 didn't trigger writing a capture
```
[gfxrecon] INFO - Settings Loader: Found option "GFXRECON_CAPTURE_TRIGGER" with value "F12"
```
This is on linux, building current `dev` branch which is at commit ef55be5720fc91db391aa9bf6f8f85558b476dd9

**Comment by dustin-lunarg** [comment]
In its current state, the hotkey capture only works on Linux with applications that use xcb surfaces.  So applications that create xlib or wayland surfaces won't work.  I believe that includes applications running on DXVK, and possibly applications using Unreal or Unity.

We should be able to add xlib surface support in the future.  Until that is done, we should add a message at surface creation to indicate hotkey capture won't be available.

**Comment by dustin-lunarg** [comment]
Adding hotkey support for the Xlib case with the existing XCB key capture code + XGetXCBConnection turned out to be pretty straightforward.  I created a branch for the changes in my fork: [dustin-lunarg:feature-xlib-hotkey](/dustin-lunarg/gfxreconstruct/tree/feature-xlib-hotkey).  This adds a new dependency, so build documentation and CMake files may need to be updated before it can merge.

**Pull request merged**